SSA optimized gensyms

vtjnash opened the pull request:
this introduces a new Symbol type as a compiler optimization for places where only a simple unnamed memory location is required. it is for use in places where the jl_varinfo_t is known by construction:

```
     int closureidx = -1
     bool isAssigned = 1
     bool isCaptured = 0
     bool isSA = 1
     bool isVolatile = 0
     bool isArgument = 0
     bool isGhost = 0/1
     bool hasGCRoot = 0/1
     bool escapes = 0
     bool usedUndef = 0
     bool used = 0/1
```

the vinfo for the Symbol should correspond to `18`

TODO:
- [x] after expanding macros, run the jlgensym renumbering pass before doing code splicing
- [x] use `(make-jlgensym)` instead of `(gensym)` in julia-syntax generation passes
- [x] build a boxed value cache for GenSym

@JeffBezanson

vtjnash committed:
wip: syntax elements of efficient indexed gensym variables

vtjnash committed:
wip: teach inference about the new GenSym type

vtjnash committed:
in flisp syntax converter, block (jlgensym) from being considered for adding as a (local)

vtjnash committed:
wip: teach codegen about the GenSym type

vtjnash committed:
wip: renumber GenSym nodes in expand to always be lowest consecutive

vtjnash committed:
wip: have inference assume the new attribute is present, rather than test-and-branch

vtjnash committed:
wip: corrections to typeinference handling of new GenSym type

vtjnash committed:
wip: more corrections and improvements to codegens handling of new GenSym type

vtjnash committed:
fix several more codegen issues

1) rename id to idx, since lldb doesn't like the name id
2) prevent back-propogation of Intrinsics.box type information
3) use SAvalue location for unboxed values as well as boxed values
4) mark a few more unboxed values (needed by 3)
5) make Type{()} a bitstype (!!!)
6) abstract the isGhost computation into type_is_ghost()
7) don't try to specsig a non-typeinferred function. emit_var may get confused by the presence of unmarked Symbols with known types (from args)
8) TODO: handle isGhost args correctly when their corresponding local variable is not a ghost

vtjnash committed:
revert making Type{()} a bitstype since this cause ambiguity. also fix emission of variables in dead code (having type Union())

vtjnash committed:
fix miscount in gensym_SAvalues creation

vtjnash committed:
trust that type-inference marked the arg type correctly

this previously could be an issue if specsig was allowed on non-type-inferred functions.

JeffBezanson commented on src/julia-syntax.scm:
```diff
@@ -132,10 +136,10 @@
 (define (sym-dot? e)
   (and (length= e 3) (eq? (car e) '|.|)
-       (symbol? (cadr e))))
+       (or (symbol? (cadr e)) (jlgensym? (cadr e)))))
```
This should be factored into a function, e.g. `symbol-like?`

vtjnash replied:
i had a feeling you would say that

Keno replied:
`symboly?`

vtjnash commented on src/julia-syntax.scm:
```diff
@@ -2984,6 +3023,7 @@ So far only the second case can actually occur.
 						(not (memq
 						      (vinfo:name v) glo))))
 			       env))
+                (gensym_types (map-int (lambda (v) 'Any) (+ (max-jlgensym (lam:body e)) 1)))
```
is there a more efficient way to create this list?

JeffBezanson replied:
Not really; this code is wonderfully lispy :)

vtjnash replied:
nvm, i found a more efficient way to create this info

vtjnash committed:
add gensym box cache and create a (symbol-like?) function for testing for (or (symbol?) (jlgensym?))

vtjnash committed:
handle GenSym objects in interpreted code, correct issues with julia-syntax.scm handling of (jlgensym) objects, and use of (make-jlgensym) where possible, disable a test that this (temporarily) broke

julia-syntax optimistically assumes many variables will be jlgensym-compatible. it is the responsiblity of branch assignment locations (such as if blocks) to handle this case appropriately, and only emit one assignment to the (jlgensym? dest) variable.

the type-inference information for GenSym variables now may be just the count of the number of GenSym variable slots present in the function

vtjnash committed:
fix more tests of GenSym operation (esp. comprehensions, and type_goto)

vtjnash committed:
fix massive performance regression caused by placing jlgensym in statement position, and cleanup of s/gensym/gensy/

vtjnash commented:
(travis failures appears to be an unrelated issues with homebrew)

it's rare i can make a PR with such a clear across-the-board performance boost:

after:
```
jameson@julia:~/julia/test/perf$ make kernel
cons                 96.266  188.395  146.604   39.080
randn_zig            37.648   37.902   37.751    0.113
gk                  158.208  404.366  248.603   93.413
sparsemul            64.683   66.159   65.245    0.590
sparsemul2           98.068  160.006  112.229   26.794
sparserange          67.659   97.540   82.707   13.952
matvec               11.285   11.452   11.351    0.066
sortperm             15.733   15.777   15.758    0.016
stockcorr          6188.298 6223.274 6206.063   13.933
bench_eu_vec        145.298  191.906  155.934   20.140
bench_eu_devec       85.300   86.161   85.519    0.363
actorgraph          894.421  962.800  924.511   29.213
laplace_vec        1864.612 1941.088 1909.872   36.892
laplace_devec       162.415  164.409  163.389    0.795
go_benchmark       1704.725 1729.694 1722.933   10.303
simplex              56.799   57.757   57.386    0.353
raytracer          2549.374 2601.338 2569.424   26.118
funarg              173.549  227.696  195.820   28.767
vectorize           254.572  297.089  266.823   17.276
splitline            75.648  117.342  108.291   18.256
json                 32.894   78.704   42.742   20.109
add1                 91.500  127.591  119.742   15.796
devec_add1           13.797   14.003   13.889    0.096
add1_logical         85.611  126.161  116.746   17.494
devec_add1_logical   16.508   16.776   16.598    0.104

fib                   0.097    0.109    0.099    0.005
parse_int             0.359    0.398    0.377    0.015
mandel                0.341    0.342    0.341    0.000
quicksort             0.716    0.804    0.736    0.038
pi_sum               55.044   55.148   55.081    0.051
rand_mat_stat        25.393   64.662   41.392   18.443
rand_mat_mul         50.424   57.288   52.698    2.648
printfd              46.103   46.433   46.228    0.145
```

before (release 0.3.5):
```
cons                119.709  210.150  173.628   33.290
randn_zig            38.245   38.271   38.259    0.013
gk                  158.591  377.580  249.697   80.515
sparsemul            65.569   67.968   66.444    1.097
sparsemul2          104.697  165.051  118.196   26.232
sparserange          68.486  103.821   84.073   15.150
matvec               10.221   11.171   10.808    0.487
sortperm             18.779   18.816   18.804    0.015
stockcorr          6270.317 6315.420 6296.297   19.654
bench_eu_vec        193.549  232.460  212.060   17.300
bench_eu_devec       89.238   90.507   89.575    0.527
actorgraph         1162.488 1711.965 1488.358  243.940
laplace_vec        2688.123 2797.466 2741.339   39.592
laplace_devec       186.556  188.872  187.950    0.883
go_benchmark       1941.020 1953.543 1948.660    5.191
simplex              59.600   59.704   59.654    0.042
raytracer          2757.393 3121.404 2894.569  156.090
funarg              214.107  281.481  241.587   35.847
vectorize           318.323  364.866  333.030   19.164
splitline            93.737  147.382  133.132   22.837
json                 33.577   81.313   43.211   21.299
add1                 99.710  151.870  138.738   22.017
devec_add1           13.622   13.758   13.701    0.055
add1_logical        105.581  155.085  143.671   21.386
devec_add1_logical   19.555   19.594   19.567    0.016

fib                   0.134    0.135    0.135    0.000
parse_int             0.398    0.659    0.500    0.103
mandel                0.334    0.353    0.338    0.008
quicksort             0.695    0.720    0.709    0.012
pi_sum               55.027   55.033   55.031    0.002
rand_mat_stat        25.236   55.845   37.485   15.779
rand_mat_mul         51.349   56.827   53.778    2.059
printfd              46.784   46.892   46.825    0.042
```

vtjnash committed:
convert scm code leading tabs to spaces

vtjnash committed:
ensure macros output unique GenSym objects also

IainNZ commented:
Here are the percentages (new times / old times), using the data in the comments above:

```
                NAME     MIN     MAX    MEAN
                cons   84.70  101.44   90.28
           randn_zig   99.62   99.63   99.66
                  gk   97.17  103.73   97.08
           sparsemul   97.66   91.41   95.67
          sparsemul2   92.54   96.43   93.21
         sparserange  106.31  111.98  108.35
              matvec  104.19   99.43  101.41
            sortperm   82.49   80.28   81.15
           stockcorr  953.27  904.32  922.60
        bench_eu_vec   96.64  105.07   98.74
      bench_eu_devec   97.42   97.68   97.48
          actorgraph  101.74  104.45  103.12
         laplace_vec   97.37  100.13   99.07
       laplace_devec   88.35   85.47   86.96
        go_benchmark   98.18   98.12   98.64
             simplex   97.03   98.19   97.81
           raytracer  100.82  100.42  100.61
              funarg  100.12   99.70   99.72
           vectorize   96.01   98.99   97.80
           splitline   99.23   98.20   98.91
                json   98.63  107.06  103.01
                add1  106.57   97.61   99.15
          devec_add1   98.95   98.21   98.52
        add1_logical   92.53   87.29   92.09
  devec_add1_logical   84.84   86.14   85.25
                 fib   72.39   81.34   73.88
           parse_int   91.35   66.22   75.55
              mandel  102.40   96.07  100.89
           quicksort  101.85  111.20  103.37
              pi_sum   99.99   99.93   99.96
       rand_mat_stat  101.30  115.80  110.38
        rand_mat_mul   98.11  106.45  100.48
             printfd   98.69   97.62   98.45
```

timholy commented:
What's the deal with the `x.ji` (large binary file) that's part of this commit?

vtjnash commented:
> Whats up with stockcorr?

good catch, i had missed that one. turns out it was a fix on base that i hadn't incorporated into my test results

> What's the deal with the x.ji (large binary file) that's part of this commit?

hm, how did that sneak back in? I rebased to get rid of it, and added a gitignore entry for it. (it's the same as sys.ji, I just like to make copies of it)

vtjnash committed:
fix win32 build

JeffBezanson commented:
Wow, is all that improvement just from having fewer symbol objects??

vtjnash commented:
it's only a few %, but that would seem to be true

JeffBezanson commented:
Quite nice to drop 2.5MB of just symbols from the system image. Yikes.

vtjnash commented:
to shrink the image a bit further, we might also try to intern strings more often. for example, we end up with a lot of cases of "unrecognized keyword argument \"". and every LineNode seems to include a full copy of the filename string.

vtjnash commented:
CI is green, so I'll go ahead and merge as soon as you are OK with it

JeffBezanson commented:
I thought the line nodes used symbols for filenames.

JeffBezanson commented on base/inference.jl:
```diff
@@ -1221,16 +1232,58 @@ function findlabel(labels, l)
+    end
+end
+
+function GenSym(sv::StaticVarInfo, typ)
```
I don't think it's good style to have a constructor mutate something. I'd call this something like `newvar!(sv, typ)`.

JeffBezanson commented on base/inference.jl:
```diff
@@ -1221,16 +1232,58 @@ function findlabel(labels, l)
+end
+function find_gensym_uses(e::ANY, uses, line)
+    if isa(e,GenSym)
+        id = (e::GenSym).id+1
```
I have a feeling these `+1`s will get annoying.

vtjnash replied:
i'm not sure what you want there. they are a bit annoying, but they save me from doing more `-1`'s in the C code. i could write an accessor function, but that seems a bit overkill. it's simply the unfortunate downside to breaking with the C-style zero-based indexing

JeffBezanson replied:
Yeah, it's not a big deal.

vtjnash commented on base/inference.jl:
```diff
@@ -1439,13 +1489,13 @@ function typeinf(linfo::LambdaStaticData,atypes::Tuple,sparams::Tuple, def, cop)
     @label typeinf_top
 
-    old_s1 = nothing
+    typegotoredo = false
```
can you check my type_goto algorithm? it's different from yours, but I couldn't entirely follow yours

JeffBezanson commented on base/inference.jl:
```diff
@@ -1479,7 +1529,20 @@ function typeinf(linfo::LambdaStaticData,atypes::Tuple,sparams::Tuple, def, cop)
+                if old===NF || !(new <: old)
+                    gensym_types[id] = tmerge(old, new)
+                    for r in gensym_uses[id]
+                        if !is(s[r],()) # s[r] === () => unreached statement
```
`s[r] === ()` doesn't mean `r` is never reached; it means it hasn't been reached *yet*.

`W` keeps track of control flow. You should not visit all places a variable is used when its type changes. You should only visit those places that the information actually reaches via control flow.

A further problem is that we will want to generalize this algorithm to derive type information from branches: inside `if isa(x,T)` we want to know `x::T`. So we should continue to track types for all variables at all program points.

vtjnash replied:
i should probably clarify that comment. we haven't reached it /yet/. since there's only one assignment site for these variables, we consider that all uses of them must have the same type information (using them undef is also disallowed). so we need to make sure to revisit any place that we've already inferred and fix their type information. Using the information that `x::T` after `isa(x,T)` would requiring creating a new GenSym in the generalized algorithm.

but we don't want to use W here to create a control flow branch to a place that may not be reached in normal control flow analysis

JeffBezanson replied:
This change is really unnecessarily complicated. It adds the `gensym_uses` data structure, plus a pass to populate it. This is redundant, since this whole algorithm is already about finding uses of everything, in a flow-sensitive way. gensym_uses finds uses in a flow-*in*sensitive way, which only works because of assumptions about how the gensyms are used.

In short I think it's simpler to handle GenSyms just how symbols are handled now: store a type at each program point, and then afterwards in `type_annotate` move the information to the var info arrays.

vtjnash replied:
part of the point of the GenSym implementation is that their information is flow-insensitive. they are only auto-generated in places where we know, a-priori, that this optimization is valid. they aren't really accessible to be created by the user (since they don't have a syntax). i could lazy-populate the `gensym_uses` vector (i should switch it to an IntSet), but I'm not sure of that value in that

vtjnash commented:
> I thought the line nodes used symbols for filenames.

i thought so too, but the text seems to be showing up too many times in `strings sys.ji`

JeffBezanson commented on src/cgutils.cpp:
```diff
@@ -1340,6 +1359,7 @@ static jl_arrayvar_t *arrayvar_for(jl_value_t *ex, jl_codectx_t *ctx)
     if (aname && ctx->arrayvars->find(aname) != ctx->arrayvars->end()) {
         return &(*ctx->arrayvars)[aname];
     }
+    //TODO: gensym case
```
Is this still to be done?

vtjnash replied:
yes, i haven't re-implemented this optimization for GenSym variables

JeffBezanson commented on src/match.scm:
```diff
@@ -190,58 +190,58 @@
-	 (expander `(lambda ,args ,body)))
+         (expander `(lambda ,args ,body)))
     `(lambda (__ex__)
        (plambda-expansion ',pat __ex__ ,expander ',args))))
```
Is this file all whitespace changes? If so could you roll them back?

vtjnash replied:
sorry, yes, i did a commit of just whitespace normalization (https://github.com/JuliaLang/julia/commit/c1ea4d66da2237d7026780d640ba0b0feeca59ec)

JeffBezanson commented on src/julia-parser.scm:
```diff
@@ -1684,216 +1684,216 @@
-		      (take-token s)
-		      '(cell1d))
-	       (let ((vex (parse-cat s #\} #t)))
+          ;; char literal
```
More all-whitespace changes? Please undo; makes the diff very hard to read.

ivarne replied:
https://github.com/JuliaLang/julia/pull/9729/files?w=1

or `git diff -w` or `git diff -b`

vtjnash replied:
i thought there used to a button to toggle it? it's interesting that the link still works. i made the commit assuming it would be easy enough to turn on ignore-whitespace

JeffBezanson commented:
How do you feel about using the GenSym approach for all local variables? (In which case I would name it something like `LocalVar`.) This would have the following advantages:

1. Every local has an index, so we can use arrays with fast lookup for all types and variable properties. We can get rid of the current clunky varinfo arrays.
1. All local names become just a matter of debug info. This would automatically fix the problem we currently have of not reporting renamed variable names correctly in error messages.
1. Type info during inference becomes just a matrix.
1. Greater uniformity. Some people have complained, pretty reasonably, that when you see a symbol in an AST it is too hard to tell what it is.

Of course, we lose the advantage that `GenSym` means SSA. However it would be quite easy/fast to check `isSSA(vinfo[x.id])`.

vtjnash commented:
it'll make an expaneded AST much harder to read, although I can see the appeal

StefanKarpinski commented:
> it'll make an expaneded AST much harder to read, although I can see the appeal

Couldn't that be dealt with by printing the AST in the more human-readable form – i.e. with readable names replaced for gensyms where there is a name to be replaced?

JeffBezanson commented:
Yes, it could be dealt with within `show`.

We know we have to fix the renamed-variables issue eventually anyway, and it seems quite unwieldy to do that by adding a symbol-to-symbol mapping to the AST, as opposed to switching to indexes for everything.

vtjnash commented:
perhaps, i'm more concerned about dealing with them in contexts where I don't have easy access to the vinfo (like `jl_`, and `emit_expr`).

fwiw, this change set also renames inlined variables in such a way that the original names now get included (i figured that was a reasonable tradeoff now to make a few more unique gensym objects here in exchange for many fewer gensym objects elsewhere)

JeffBezanson commented:
`jl_` is a problem, I see that. Usual just-for-debugging arguments apply :) But surely `emit_expr` has easy access to all var info.

Anyway we need to think really hard about making our ASTs as simple and efficient as possible. They are already too hard to work with (main problem), and account for too much memory traffic.

vtjnash commented:
i mostly meant `jl_` usage /in/ `emit_expr`. although `break set -n abc -c 'sym == symbol("hi")'` is also rather handy (in lldb syntax)

vtjnash commented:
> Anyway we need to think really hard about making our ASTs as simple and efficient as possible. They are already too hard to work with (main problem), and account for too much memory traffic.

what if we were to instead split the GenSym into positive/negative? a negative value would indicate an auto-generated ssa-gensym and a positive value would indicate a normal variable (with all info in the vinfo array). part of that is tricky because these GenSyms currently can only exist in contexts where they cannot have names. so they are disallowed as argument names, closures, and let variables, for example.  allowing them in those context would require much more work, and I would like merge this and go back to fixing ccall (leaving your proposed secondary transform as future work).

JeffBezanson commented:
Yes, I agree the further changes can be done later after this PR.

Closed variables definitely need a different approach. It would be better to represent them in the AST as `(closedvar i)`, and represent assignments to them with `setfield`. Or we could do an even more radical rewrite of closures using `call` overloading.

I question whether it really is so important to be able to instantly distinguish SSA vars. To me it looks like the main benefits are fewer symbol objects, and possibly more uniform variable metadata in the future. Using GenSym for all locals, there would be a UInt8 array for variable properties. The savings for knowing that many vars are SSA is one byte per variable, which is not much. Maybe it will help that most SSA vars are added later, and we can avoid growing the vinfo and name mapping arrays for vars that don't need them.

vtjnash commented:
> Maybe it will help that most SSA vars are added later, and we can avoid growing the vinfo and name mapping arrays for vars that don't need them.

that's mostly the help. but it also should help save on type inference effort, by not bothering to track their dataflow, and not bothering to create SymbolNode's everywhere (although perhaps those could also go away for Symbols with a constant type, if variables were just indices)

vtjnash commented:
that might even be a good exercise to do when specializing all Expr objects (and making Expr abstract)

JeffBezanson commented:
I would rather track their dataflow than need to insert extra variables and assignments to get type info from branches. Plus, the type info obtained this way must be strictly a refinement of types we already know, so the overall type stored for an SSA var is still correct.

Also my point was that even if we use GenSyms for everything, we can still avoid storing some metadata for the late-added SSA ones.

vtjnash commented:
> I would rather track their dataflow than need to insert extra variables and assignments to get type info from branches. Plus, the type info obtained this way must be strictly a refinement of types we already know, so the overall type stored for an SSA var is still correct.

true, but i think codegen often needs the extra variable to make much use of the information (to know where the unboxed value is stored).

but i don't really intend for these to be used in contexts where a typeassert is meaningful. i don't see a reason the compiler would be inserting that. they should not fail if they got forced into that situation somehow, but a typeassert seems to imply the variable has a name. GenSym variables do not.

vtjnash committed:
cleanup, per Jeff's comments

JeffBezanson commented:
That is all true, but I still maintain that it would be simpler to use indexes for all locals and treat all variables uniformly in dataflow analysis. I think there will be less to worry about that way, and if it makes ASTs easier to work with it's a win.

vtjnash commented:
i don't disagree, it's just not the optimization i was attempting here

JeffBezanson commented:
In fact, why not remove ~50 lines of code from this PR by removing `gensym_uses` and just playing along with the existing dataflow algorithm?

I haven't looked at the new type_goto algorithm yet, but why was it needed?

vtjnash commented:
> In fact, why not remove ~50 lines of code from this PR by removing gensym_uses and just playing along with the existing dataflow algorithm?

i could, but then i would need to add ~50 lines so that the existing dataflow algorithm knows what to do when it encounters a GenSym. this way, i was hoping to avoid some work with testing for tchanged

the previous `type_goto` algorithm seemed to be partially unaware of dataflow, and partially aware of it. i'm not sure if it was entirely correct, although I also don't know of any cases where it would fail. the updated algorithm only works with dataflow-independent objects (aka GenSym's), but I think it does less work overall. i thought the old algorithm may have worked mostly because the objects it had to deal with were in fact SSA (which is why it was valid to copy their type information into the first line)

in particular, i couldn't understand the purpose of these lines https://github.com/JuliaLang/julia/blob/adc1c83cc3ee599604998c5d56991826ca2acc9f/base/inference.jl#L1516-L1518

vtjnash commented:
note, if we assume that the test suite is a test of compile time, we see an improvement on every test in my measured subset from this PR:

after:
```
jameson@julia:~/julia/test$ make core arrayops subarray
    JULIA test/core
     * core                 in  11.26 seconds
    SUCCESS
    JULIA test/arrayops
     * arrayops             in  35.67 seconds
    SUCCESS
    JULIA test/subarray
     * subarray             in  66.37 seconds
    SUCCESS
jameson@julia:~/julia/test$ make linalg
    JULIA test/linalg
	From worker 6:	     * linalg/lapack        in  10.99 seconds
	From worker 6:	     * linalg/cholmod       in   0.21 seconds
	From worker 8:	     * linalg/tridiag       in  13.26 seconds
	From worker 8:	     * linalg/givens        in   3.60 seconds
	From worker 6:	     * linalg/umfpack       in   6.48 seconds
	From worker 9:	     * linalg/pinv          in  18.71 seconds
	From worker 4:	     * linalg3              in  30.70 seconds
	From worker 5:	     * linalg4              in  33.37 seconds
	From worker 3:	     * linalg2              in  82.83 seconds
	From worker 2:	     * linalg1              in  84.17 seconds
	From worker 7:	     * linalg/triangular    in 112.19 seconds
    SUCCESS
```

before (master adc1c83):
```
jameson@julia:~/julia-reference$ cd test/
jameson@julia:~/julia-reference/test$ make core arrayops subarray
    JULIA test/core
     * core                 in  11.46 seconds
    SUCCESS
    JULIA test/arrayops
     * arrayops             in  39.88 seconds
    SUCCESS
    JULIA test/subarray
     * subarray             in  70.74 seconds
    SUCCESS
jameson@julia:~/julia-reference/test$ make linalg
    JULIA test/linalg
	From worker 6:	     * linalg/lapack        in  13.08 seconds
	From worker 6:	     * linalg/cholmod       in   0.24 seconds
	From worker 8:	     * linalg/tridiag       in  14.85 seconds
	From worker 8:	     * linalg/givens        in   3.50 seconds
	From worker 9:	     * linalg/pinv          in  19.76 seconds
	From worker 6:	     * linalg/umfpack       in   6.89 seconds
	From worker 4:	     * linalg3              in  34.84 seconds
	From worker 5:	     * linalg4              in  40.50 seconds
	From worker 3:	     * linalg2              in  85.28 seconds
	From worker 2:	     * linalg1              in  93.01 seconds
	From worker 7:	     * linalg/triangular    in 129.54 seconds
    SUCCESS
```

ViralBShah commented:
Looks like the linalg tests do not benefit and even slow down slightly. I always wish they get faster.

JeffBezanson commented:
Look again; the "after" numbers are listed first. Looks like almost all the linalg tests are significantly faster. Only a couple are a hair slower; probably not statistically significant.

JeffBezanson commented:
Anyway @vtjnash back to type_goto. Fortunately there are tests for all the relevant issues, and they pass, so hopefully we're in good shape. Your algorithm looks at least as good as mine AFAICT.

vtjnash commented:
OK. my algorithm avoids computing `tchanged` as often, thus potentially using a slightly different termination condition. but otherwise, it isn't too much of a departure so it wasn't hard to make the test pass

vtjnash commented:
unfortunately, one of the torture test i was hoping would be greatly improved by this PR (https://gist.github.com/vtjnash/27ff622b2cbed22b51dc) seems to have been not affected much / in the wrong direction:

after:
```
julia> @time include("../slow_performance.jl")
elapsed time: 0.024892235 seconds (333 kB allocated)
elapsed time: 3.61572336 seconds (7 MB allocated, 1.15% gc time in 1 pauses with 0 full sweep)

julia> @time show(t)
elapsed time: 24.334687017 seconds (2050 MB allocated, 10.09% gc time in 26 pauses with 13 full sweep)
2-element Array{T1Type,1}:
 
 

julia> 
```

before:
```
julia> @time include("../slow_performance.jl")
elapsed time: 0.023578148 seconds (334 kB allocated)
elapsed time: 3.447290547 seconds (7 MB allocated, 1.16% gc time in 1 pauses with 0 full sweep)

julia> @time show(t)
elapsed time: 50.349472299 seconds (2153 MB allocated, 5.71% gc time in 31 pauses with 13 full sweep)
2-element Array{T1Type,1}:
 
 

julia>
```

JeffBezanson commented:
The first time seems to be within 10%. Isn't the second one 2x faster with the new code?

ViralBShah commented:
Yes - I misread the timings. Sorry for the noise.

vtjnash commented:
> The first time seems to be within 10%. Isn't the second one 2x faster with the new code?

oh, i was just looking at parse (actually expand) times, not the type-inference pass. yes, it seems it did help that a bit.
